File: Assets/MLKitWrapper/MLKitWrapper.cs
//#define DUMMY_MODE
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class MLKitWrapper : MonoBehaviour
{
	public const int InvalidRequestId = int.MinValue;

	[SerializeField] int parallelCount = 2;

	[System.Serializable]
	public class Text
	{
		public string text;
		public TextBlock[] textBlocks;
		public int requestId;
	}

	[System.Serializable]
	public class TextBlock
	{
		public Rect boundingBox;
		public Point[] cornerPoints;
		public Line[] lines;
		public string recognizedLanguage;
		public string text;
	}

	[System.Serializable]
	public class Line
	{
		public Rect boundingBox;
		public Point[] cornerPoints;
		public Element[] elements;
		public string recognizedLanguage;
		public string text;
	}

	[System.Serializable]
	public class Element
	{
		public Rect boundingBox;
		public Point[] cornerPoints;
		public string recognizedLanguage;
		public string text;
	}

	[System.Serializable]
	public class Rect
	{
		public int bottom;
		public int left;
		public int right;
		public int top;
	}

	[System.Serializable]
	public class Point
	{
		public int x;
		public int y;
	}

	public bool Implemented
	{ 
		get
		{
			var ret = false;
#if UNITY_EDITOR // Editorでは無効

#elif UNITY_ANDROID
			ret = true;
#endif			

#if DUMMY_MODE
			ret = true;
#endif
			return ret;
		}
	}

	// キューが埋まっているか
	public bool IsFull()
	{
		EnsureInitialize();
		return (FindEmptyRequestIndex() < 0);
	}

	// IDがInvalidなら何でもいいから返す
	public Text GetResult(bool resetOnGet, int requestId = InvalidRequestId)
	{
		EnsureInitialize();

		Text ret = null;
		foreach (var request in requests)
		{
			if (request.IsDone())
			{
				if ((request.requestId == requestId) || requestId == InvalidRequestId)
				{
					ret = request.output.text;
					ret.requestId = request.requestId;
					if (resetOnGet)
					{
Debug.Log("GetResult : " + resetOnGet + " " + request.requestId);
						request.Reset();
					}
					break;
				}
			}
		}
		return ret;
	}

	// 戻り値がInvalidRequestIdなら失敗している。一杯の時に勝手に止めたりはしない。
	public int RecognizeText(int width, int height, IReadOnlyList<Color32> pixels)
	{
		EnsureInitialize();

		Request newRequest = null;
		var newRequestId = InvalidRequestId;
		if (requests.Count < parallelCount)
		{
			newRequest = new Request();
			requests.Add(newRequest);
Debug.Log("Enqueue " + requests.Count);
		}
		else
		{
			var vacantIndex = FindEmptyRequestIndex();
			if (vacantIndex >= 0)
			{
				newRequest = requests[vacantIndex];
			}
		}

		if (newRequest == null)
		{
			return newRequestId;
		}

//Debug.Log("MLKit Recognize: " + width + "x" + height + " pixelCount=" + pixels.Count + " id=" + nextRequestId);

		var input = new Input();
		input.width = width;
		input.height = height;
		input.pixels = new int[pixels.Count];
		input.requestId = nextRequestId;
		newRequest.requestId = nextRequestId;

		newRequestId = nextRequestId;
		// Y反転しながら送る
		for (var y = 0; y < height; y++)
		{
			var dstY = height - y - 1;
			for (var x = 0; x < width; x++)
			{
				var src = pixels[(y * width) + x];
				input.pixels[(dstY * width) + x] = (0xff << 24) | (src.r << 16) | (src.g << 8) | src.b;
			}
		}

		try
		{
			// jsonにすると遅いので、stringにバイナリデータ詰める

			string serializedInput = null;
#if true
			serializedInput = input.Serialize();
#else
			var json = JsonUtility.ToJson(input, true);
//Debug.Log("MLKit input jsonSize=" + json.Length + " req=" + Requested);
#if UNITY_EDITOR
System.IO.File.WriteAllText("mlkitInput.json", json);
#endif
#endif
#if DUMMY_MODE
			StartCoroutine(CoOnCompleteDummy(nextRequestId));
#else
			javaClass.CallStatic("recognizeText", gameObject.name, serializedInput);
#endif
		}
		catch (System.Exception e)
		{
			Debug.LogException(e);
		}

		nextRequestId++;
		if (nextRequestId == int.MinValue) // 来ないと思うが巻き戻った時のための対処
		{
			nextRequestId++;
		}

		return newRequestId;
	}

	// non public -----
	[System.Serializable]
	class Input
	{
		// Jsonだと遅いのでバイナリでString内にSerializeする
		public string Serialize()
		{
			var sb = new System.Text.StringBuilder();
			Append(sb, requestId);
			Append(sb, width);
			Append(sb, height);
			var n = width * height;
			Debug.Assert(n == pixels.Length);
			for (var i = 0; i < n; i++)
			{
				Append(sb, pixels[i]);
			}
			return sb.ToString();
		}
		public int requestId;
		public int width;
		public int height;
		public int[] pixels;

		// non public ----
		void Append(System.Text.StringBuilder sb, int v)
		{
			sb.Append((char)(v & 0xffff)); //Low
			v >>= 16;
			sb.Append((char)(v & 0xffff)); //High
		}
	}
	
	[System.Serializable]
	class Output
	{
		public int requestId;
		public Text text;
		public string errorMessage;
	}

	class Request
	{
		public Request()
		{
			Reset();
		}

		public void SetResult(Output output)
		{
			this.output = output;
		}

		public bool IsDone()
		{
			var ret = false;
			if (requestId != InvalidRequestId)
			{
				if (output != null)
				{
					ret = true;
				}
			}
			return ret;
		}

		public void Reset()
		{
			requestId = InvalidRequestId;
			output = null;
		}
		public Output output;
		public int requestId;
	}

	AndroidJavaClass javaClass;
	List<Request> requests;
	int nextRequestId; // 20億使い果たしたらバグります

	int FindEmptyRequestIndex()
	{
		var ret = int.MinValue;
		// 数が足りなければ足してそれ返す
		if (requests.Count < parallelCount)
		{
			var request = new Request();
			ret = requests.Count;
			requests.Add(request);
		}
		else
		{
			for (var i = 0; i < requests.Count; i++)
			{
				var request = requests[i];
				if (request.requestId == InvalidRequestId)
				{
					ret = i;
					break;
				}
			}
		}
		return ret;
	}


	Request FindRequest(int requestId)
	{
		var index = FindRequestIndex(requestId);
		Request ret = null;
		if (index >= 0)
		{
			ret = requests[index];
		}
		return ret;
	}

	int FindRequestIndex(int requestId)
	{
//Debug.Log("Find id=" + requestId + " reqs=" + requests.Count);
		var ret = int.MinValue;
		for (var i = 0; i < requests.Count; i++)
		{
			var request = requests[i];
//Debug.Log("\t" + i + " " + request.requestId);
			if (request.requestId == requestId)
			{
				ret = i;
				break;
			}
		}
		return ret;
	}

	void EnsureInitialize()
	{
		if (requests == null)
		{
			requests = new List<Request>();
		}

		try
		{
			if (javaClass == null)
			{
				javaClass = new AndroidJavaClass("com.hirasho.MLKitWrapper.MLKitWrapper");
			}
		}
		catch (System.Exception e)
		{
			Debug.LogException(e);
		}
	}
 
	// JAVAから呼ばれるので参照はない
	void OnComplete(string outputJson) 
	{
Debug.Log("MLKitWrapper.OnComplete: json\n" + outputJson);
		Output output = null;
		try
		{
			output = JsonUtility.FromJson<Output>(outputJson);
		}
		catch (System.Exception e)
		{
			Debug.LogException(e);
		}
		
		if (output != null)
		{
			var request = FindRequest(output.requestId);
			if (request != null)
			{
				request.SetResult(output);
			}
		}
	}

	IEnumerator CoOnCompleteDummy(int requestId)
	{
		yield return new WaitForSeconds(0.25f);
		Debug.Log("MLKitWrapper.OnComplete(DUMMY)");

		var request = FindRequest(requestId);
		if (request != null)
		{
			var boundingBox = new Rect() { bottom = 432 - 8, left = 8, right = 768 - 8, top = 8 };

			var cornerPoints = new Point[4];
			cornerPoints[0] = new Point() { x = 8, y = 8 };
			cornerPoints[1] = new Point() { x = 768 - 8, y = 8 };
			cornerPoints[2] = new Point() { x = 768 - 8, y = 432 - 8 };
			cornerPoints[3] = new Point() { x = 8, y = 432 - 8 };

			var text = new Text();
			text.text = "HOGE";
			text.textBlocks = new TextBlock[1];

			var textBlock = new TextBlock();
			textBlock.boundingBox = boundingBox;
			textBlock.cornerPoints = cornerPoints;
			textBlock.recognizedLanguage = "English";
			textBlock.text = "HOGE";
			textBlock.lines = new Line[1];
			text.textBlocks[0] = textBlock;

			var line = new Line();
			line.boundingBox = boundingBox;
			line.cornerPoints = cornerPoints;
			line.recognizedLanguage = "English";
			line.text = "HOGE";
			line.elements = new Element[1];
			textBlock.lines[0] = line;

			var element = new Element();
			element.boundingBox = boundingBox;
			element.cornerPoints = cornerPoints;
			element.recognizedLanguage = "English";
			element.text = "HOGE";
			line.elements[0] = element;

			var output = new Output();
			output.text = text;

			request.SetResult(output);
		}
	}
}

File: Assets/MLKitWrapper/MLKitWrapper.java
package com.hirasho.MLKitWrapper;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.text.Text;
import com.google.mlkit.vision.text.TextRecognition;
import com.google.mlkit.vision.text.TextRecognizer;
import com.google.mlkit.vision.text.latin.TextRecognizerOptions;
import com.unity3d.player.UnityPlayer;
import com.google.gson.Gson;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.graphics.Point;
import java.util.List;
import android.util.Log;

public class MLKitWrapper 
{
	static public class Input
	{
		static public Input deserialize(String serialized)
		{
			Input ret = new Input();
			ret.requestId = readInt(serialized, 0);
			ret.width = readInt(serialized, 2);
			ret.height = readInt(serialized, 4);
			int n = ret.width * ret.height;
			ret.pixels = new int[n];
			for (int i = 0; i < n; i++)
			{
				ret.pixels[i] = readInt(serialized, 6 + (i * 2));
			}
			return ret;
		}

		public int requestId;
		public int width;
		public int height;
		public int[] pixels;

		static int readInt(String serialized, int pos)
		{
			int ret = serialized.charAt(pos);
			ret |= serialized.charAt(pos + 1) << 16;
			return ret;
		}
	}
	
	static public class Output
	{
		public int requestId;
		public TextData text;
		public String errorMessage;
	}

	// 以下~Dataで終わるクラスはMLKitのクラスを写したもの
	static class TextData
	{
		public String text;
		public TextBlockData[] textBlocks;
	}

	static class TextBlockData
	{
		public RectData boundingBox;
		public PointData[] cornerPoints;
		public LineData[] lines;
		public String recognizedLanguage;
		public String text;
	}

	static class LineData
	{
		public RectData boundingBox;
		public PointData[] cornerPoints;
		public ElementData[] elements;
		public String recognizedLanguage;
		public String text;
	}

	static class ElementData
	{
		public RectData boundingBox;
		public PointData[] cornerPoints;
		public String recognizedLanguage;
		public String text;
	}

	static class RectData
	{
		public String toString()
		{
			return "(" + left + "," + top + ") - (" + right + "," + bottom + ")";
		}
		public int bottom;
		public int left;
		public int right;
		public int top;
	}

	static class PointData
	{
		public int x;
		public int y;
	}
	
	static public void recognizeText(String receiverGameObjectName, String serializedInput) 
	{
		TextRecognizer recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS);

//		Gson gson = new Gson();
//		Input input = gson.fromJson(serializedInput, Input.class);
		Input input = Input.deserialize(serializedInput);
Log.d("MLKit", "input " + Integer.toString(input.width) + "x" + Integer.toString(input.height) + " pixelCount=" + Integer.toString(input.pixels.length) + " id=" + Integer.toString(input.requestId));

		Bitmap bitmap = Bitmap.createBitmap(input.pixels, input.width, input.height, Bitmap.Config.ARGB_8888);
		if (bitmap == null)
		{
			Log.e("MLKit", "Bitmap.createBitmap failed.");
		}
		else
		{
			InputImage inputImage = InputImage.fromBitmap(bitmap, 0);
			if (inputImage == null)
			{
				Log.e("MLKit", "InputImage.fromBitmap failed.");
			}
			else
			{
				Output output = new Output();
				output.requestId = input.requestId;

				OnSuccess onSuccess = new OnSuccess(receiverGameObjectName, output);
				OnFailure onFailure = new OnFailure(receiverGameObjectName, output);
				Task<Text> result = recognizer.process(inputImage).
					addOnSuccessListener(onSuccess).
					addOnFailureListener(onFailure);
			}
		}
	}

	static class OnSuccess implements OnSuccessListener<Text>
	{
		public OnSuccess(String receiverGameObjectName, Output output)
		{
			this.receiverGameObjectName = receiverGameObjectName;
			this.output = output;
		}

		@Override public void onSuccess(Text text) 
		{
			if (text == null)
			{
				Log.e("MLKit", "OnSuccess.onSuccess text is null.");
			}
			else
			{
				TextData textData = new TextData();
				try
				{
					output.text = parseText(text);
				}
				catch (Exception e)
				{
					Log.e("MLKit", "OnSuccess parseText throw exception: " + e.getMessage());
				}

				if (output.text != null)
				{
					Gson gson = new Gson();
					String outputJson = gson.toJson(output);
					UnityPlayer.UnitySendMessage(receiverGameObjectName, "OnComplete", outputJson);
				}
			}
		}
		// non public ----
		String receiverGameObjectName;
		Output output;

		TextData parseText(Text src)
		{
			TextData ret = new TextData();
			ret.text = src.getText();

			List<Text.TextBlock> srcTextBlocks = src.getTextBlocks();
			if (srcTextBlocks != null)
			{
				ret.textBlocks = new TextBlockData[srcTextBlocks.size()];
Log.d("MLKit", "parseText " + ret.text + " " + ret.textBlocks.length);
				for (int i = 0; i < srcTextBlocks.size(); i++)
				{
					TextBlockData block = parseTextBlock(srcTextBlocks.get(i));
					ret.textBlocks[i] = block;
				}
			}
			else
			{
				Log.w("MLKit", "OnSuccess.parseText getTextBlocks returned null.");
			}
			return ret;
		}

		TextBlockData parseTextBlock(Text.TextBlock src)
		{
			TextBlockData ret = new TextBlockData();
			ret.boundingBox = parseRect(src.getBoundingBox());
			ret.cornerPoints = parsePoints(src.getCornerPoints());
			ret.recognizedLanguage = src.getRecognizedLanguage();
			ret.text = src.getText();

			List<Text.Line> srcLines = src.getLines();
			if (srcLines != null)
			{
				ret.lines = new LineData[srcLines.size()];
Log.d("MLKit", "parseTextBlock " + ret.text + " " + ret.lines.length + " " + ret.boundingBox);
				for (int i = 0; i < srcLines.size(); i++)
				{
					LineData line = parseLine(srcLines.get(i));
					ret.lines[i] = line;
				}
			}
			else
			{
				Log.w("MLKit", "OnSuccess.parseTextBlock getLines returned null.");
			}
			return ret;
		}

		LineData parseLine(Text.Line src)
		{
			LineData ret = new LineData();
			ret.boundingBox = parseRect(src.getBoundingBox());
			ret.cornerPoints = parsePoints(src.getCornerPoints());
			ret.recognizedLanguage = src.getRecognizedLanguage();
			ret.text = src.getText();

			List<Text.Element> srcElements = src.getElements();
			if (srcElements != null)
			{
				ret.elements = new ElementData[srcElements.size()];
Log.d("MLKit", "parseLine " + ret.text + " " + ret.elements.length + " " + ret.boundingBox);
				for (int i = 0; i < srcElements.size(); i++)
				{
					ElementData element = parseElement(srcElements.get(i));
					ret.elements[i] = element;
				}
			}
			else
			{
				Log.w("MLKit", "OnSuccess.parseLine getElements returned null.");
			}

			return ret;
		}

		ElementData parseElement(Text.Element src)
		{
			ElementData ret = new ElementData();
			ret.boundingBox = parseRect(src.getBoundingBox());
			ret.cornerPoints = parsePoints(src.getCornerPoints());
			ret.recognizedLanguage = src.getRecognizedLanguage();
			ret.text = src.getText();
Log.d("MLKit", "parseElement " + ret.text + " " + ret.boundingBox);
			return ret;
		}

		RectData parseRect(Rect src)
		{
			RectData ret = new RectData();
			ret.bottom = src.bottom;
			ret.left = src.left;
			ret.right = src.right;
			ret.top = src.top;
			return ret;
		}

		PointData parsePoint(Point src)
		{
			PointData ret = new PointData();
			ret.x = src.x;
			ret.y = src.y;
			return ret;
		}

		PointData[] parsePoints(Point[] src)
		{
			PointData[] ret = new PointData[src.length];
			for (int i = 0; i < src.length; i++)
			{
				ret[i] = parsePoint(src[i]);
			}
			return ret;
		}
	}

	static class OnFailure implements OnFailureListener
	{
		public OnFailure(String receiverGameObjectName, Output output)
		{
			this.receiverGameObjectName = receiverGameObjectName;
			this.output = output;
		}

		@Override public void onFailure(Exception e) 
		{
			output.errorMessage = e.getMessage();
Log.d("MLKit", "onFailure " + output.errorMessage);

			Gson gson = new Gson();
			String outputJson = gson.toJson(output);
			UnityPlayer.UnitySendMessage(receiverGameObjectName, "OnComplete", outputJson);
		}
		// non public ----
		String receiverGameObjectName;
		Output output;
	}

	static public int incrementTest(int a) 
	{
		return a + 1;
	}
}

